subscriptions/providers/apple_in_app/app_store.py

```diff
@@ -0,0 +1,261 @@
+import base64
+import dataclasses
+import datetime
+import enum
+import pathlib
+from typing import (
+    Any,
+    Optional,
+)
+
+import jwt
+from OpenSSL import crypto
+from cryptography.hazmat.primitives.serialization import Encoding
+from cryptography.x509 import load_der_x509_certificate
+
+from .api import (
+    AppleEnvironment,
+    datetime_from_ms_timestamp,
+)
+
+CACHED_APPLE_ROOT_CERT: Optional[crypto.X509] = None
+
+
+class AppleAppStoreError(Exception):
+    pass
+
+
+class ConfigurationError(AppleAppStoreError):
+    pass
+
+
+class PayloadValidationError(AppleAppStoreError):
+    pass
+
+
+def load_certificate_from_bytes(certificate_data: bytes) -> crypto.X509:
+    basic_cert = load_der_x509_certificate(certificate_data)
+    return crypto.X509.from_cryptography(basic_cert)
+
+
+def load_certificate_from_x5c(x5c_entry: str) -> crypto.X509:
+    # Each string in the array is a base64-encoded (not base64url-encoded) DER [ITU.X690.2008] PKIX certificate value.
+    certificate_data = base64.b64decode(x5c_entry)
+    return load_certificate_from_bytes(certificate_data)
+
+
+def setup_original_apple_certificate(certificate_path: str) -> None:
+    global CACHED_APPLE_ROOT_CERT
+
+    if CACHED_APPLE_ROOT_CERT is None:
+        cert_path = pathlib.Path(certificate_path)
+        if not cert_path.exists() or not cert_path.is_file():
+            raise ConfigurationError('No root certificate for Apple provided. Check Django configuration settings.')
+
+        CACHED_APPLE_ROOT_CERT = load_certificate_from_bytes(cert_path.read_bytes())
+
+
+def get_original_apple_certificate() -> crypto.X509:
+    global CACHED_APPLE_ROOT_CERT
+    if CACHED_APPLE_ROOT_CERT is None:
+        raise ConfigurationError('Certificate was not set up properly. '
+                                 'Call setup_original_apple_certificate before using it.')
+    return CACHED_APPLE_ROOT_CERT
+
+
+def are_certificates_identical(cert_1: crypto.X509, cert_2: crypto.X509) -> bool:
+    # There are a few ways to compare these, but checking their binary representation seems good enough.
+    return cert_1.to_cryptography().public_bytes(Encoding.DER) == cert_2.to_cryptography().public_bytes(Encoding.DER)
+
+
+def validate_and_fetch_apple_signed_payload(signed_payload: str) -> dict[str, Any]:
```

Comment:
That's a good one.

I tried searching for libraries that handle this properly (like, perform a correct validation and stuff) and failed. And if you don't really care you can just decode the payload without any form of validation. There were gists in languages like `go`, and this code is result of all the solutions converted to Python (and confirmed, more or less, with tests).

As for JWT – there are libraries for it (I even use one of them, the one with most stars and least issues), but they don't support the `x5c` (certificate chain) header. Checking whether it's ok or not still falls on the developer. And it's not that scary once you know everything:
- `x5c` contains a list of base64 encoded DER certificates;
- there should be 3 of them, the last being the root from Apple;
- the public key from the first one is used to sign the payload.
The problem is that "validating certificate chain" doesn't seem to be the simplest task, even in 2022. I could potentially spend more time checking whether some library (different than `cryptography` and `OpenSSL`) can handle it properly, but this flow presented here "was confirmed" to be fine by someone smarter than me in that regard.

And as you wrote about using nudge from google to pull the payment information – this also can be done on the Apple side and I didn't think about it. The process would be slightly simpler (as JWT token is still needed https://developer.apple.com/documentation/appstoreserverapi/creating_api_keys_to_use_with_the_app_store_server_api and https://developer.apple.com/documentation/appstoreserverapi/generating_tokens_for_api_requests), but no `x5c` chain verification would be required. But it relies on another request being made against the Apple API while here we have all the data in our hand already.

Is it too much? Should I abstract it away even further?

Reply:
I also thought about https, so if you know which servers apple uses to send the notifications, you could just check that request came from those and not care about anything else - because https does certificate job for you :) I haven't found any list of servers but I haven't searched for it too thoroughly.

Tbh I don't know the best solution for this. As you already wrote it and it works, it would be weird to try anything else unless that "something else" is easy to integrate and use. The main concern is that touching cryptography is not something we really want (that's my opinion, not RT's policy) :) Maybe ask on #python about all this? High chances that Pawel will ask to opensource this jwt-crypto stuff

Reply:
I also haven't found any servers that would be the source of Apple queries. And their documentation explains the process that should be undertaken rather thoroughly and they mention no IPs, DNS names nor anything else that should/could be used to discriminate the input. Also, we're using the same endpoint for client communication, so that any kind of verification would have to happen on application level...

I'll ask about it on `#python`. Thanks for the tip!